File: server/app/routers/deep_analysis.py
"""Deep Analysis router — 5 diagnostic tools backed by SIF MCP + AI agents.

Tools:
  1. keyword    — 关键词竞争分析 (market_get_keyword_competition)
  2. competitor — 竞品反查 (market_get_asin_keyword_signals)
  3. traffic    — 流量异动诊断 (analyze_traffic_anomaly)
  4. reviews    — 评论聚类 (AI agent analysis)
  5. listing    — Listing 批量改写 (listing service)
"""
from __future__ import annotations

import asyncio
import json
import time
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException
from fastapi.responses import StreamingResponse
from pydantic import BaseModel, Field

from app.core.security import require_user
from app.services import sif_service

router = APIRouter(dependencies=[Depends(require_user)])


# ── Request / Response models ─────────────────────────────────────────────

class KeywordReq(BaseModel):
    keyword: str = Field(..., min_length=1, description="关键词")
    country: str = Field("US", description="站点代码")
    asin: str = Field("", description="可选：对标 ASIN")


class CompetitorReq(BaseModel):
    asin: str = Field(..., min_length=1, description="目标 ASIN")
    country: str = Field("US", description="站点代码")
    time_type: str = Field("lately", description="时间类型: lately/week/month")
    time_value: str = Field("7", description="时间值")


class TrafficReq(BaseModel):
    asin: str = Field(..., min_length=1, description="目标 ASIN")
    country: str = Field("US", description="站点代码")


class ReviewsReq(BaseModel):
    asin: str = Field(..., min_length=1, description="目标 ASIN")
    country: str = Field("US", description="站点代码")
    marketplace: str = Field("US", description="站点（兼容字段）")


class ListingRewriteReq(BaseModel):
    asins: list[str] = Field(..., min_length=1, description="ASIN 列表")
    marketplace: str = Field("US", description="站点")
    fields: list[str] = Field(
        default=["title", "bullets"],
        description="要改写的字段: title, bullets, description, qa"
    )
    style: str = Field("professional", description="改写风格")


# ── SSE helpers ───────────────────────────────────────────────────────────

def _sse(event: dict) -> str:
    return f"data: {json.dumps(event, ensure_ascii=False)}\n\n"


_SSE_HEARTBEAT = ":hb\n\n"


# ── 1. 关键词竞争分析 ────────────────────────────────────────────────────

@router.post("/keyword")
async def keyword_competition(req: KeywordReq) -> dict:
    """关键词竞争格局分析，返回 Top ASIN、集中度、可进入性评估。"""
    try:
        data = await sif_service.keyword_competition(
            req.keyword, req.country, req.asin
        )
        return {"ok": True, "data": data}
    except Exception as exc:
        raise HTTPException(502, f"SIF MCP 调用失败: {exc}")


# ── 2. 竞品反查 ──────────────────────────────────────────────────────────

@router.post("/competitor")
async def competitor_lookup(req: CompetitorReq) -> dict:
    """竞品 ASIN 流量词反查，返回关键词信号、排名演变、健康分级。"""
    try:
        data = await sif_service.competitor_keyword_signals(
            req.asin, req.country, req.time_type, req.time_value
        )
        return {"ok": True, "data": data}
    except Exception as exc:
        raise HTTPException(502, f"SIF MCP 调用失败: {exc}")


# ── 3. 流量异动诊断 ─────────────────────────────────────────────────────

@router.post("/traffic")
async def traffic_diagnosis(req: TrafficReq) -> dict:
    """ASIN 流量下跌根因分析，自动识别异常窗口并逐层拆因。"""
    try:
        data = await sif_service.traffic_anomaly(req.asin, req.country)
        return {"ok": True, "data": data}
    except Exception as exc:
        raise HTTPException(502, f"SIF MCP 调用失败: {exc}")


# ── 4. 评论聚类 ──────────────────────────────────────────────────────────

@router.post("/reviews")
async def review_clustering(req: ReviewsReq) -> StreamingResponse:
    """评论聚类分析 — 差评差异化成因识别与修复建议。SSE 流式返回。"""
    from app.services import ai_synthesis_service

    prompt = f"""你是一位 Amazon 产品评论分析专家。
请对 ASIN {req.asin}（站点 {req.country}）的评论进行聚类分析：

1. 收集该 ASIN 的评论数据（使用 sorftime product_reviews 工具）
2. 对差评（1-3星）进行主题聚类，识别 Top 5 差评原因
3. 对好评（4-5星）进行主题聚类，识别 Top 3 产品优势
4. 针对每个差评原因，给出具体的改进建议（Listing 文案优化 / 产品改进 / 售后策略）
5. 输出格式：
   - 差评聚类表格：主题 | 占比 | 典型评论摘要 | 改进建议
   - 好评聚类表格：主题 | 占比 | 可利用的卖点
   - 优先级排序的行动建议

请用中文输出，数据尽量用表格呈现。"""

    async def generator():
        start = time.time()
        yield _sse({"type": "phase", "phase": "collecting"})
        try:
            async for prov, chunk in ai_synthesis_service.synthesize_native(
                "asin", req.asin, req.country
            ):
                if prov == "_attempt":
                    yield _sse({"type": "attempt", "provider": chunk})
                elif prov == "error":
                    yield _sse({"type": "error", "detail": chunk})
                    return
                else:
                    yield _sse({"type": "token", "text": chunk, "provider": prov})
            elapsed = round(time.time() - start, 1)
            yield _sse({"type": "done", "provider": "hermes", "elapsed_s": elapsed})
        except Exception as exc:
            yield _sse({"type": "error", "detail": str(exc)})

    return StreamingResponse(
        generator(),
        media_type="text/event-stream",
        headers={"Cache-Control": "no-cache", "X-Accel-Buffering": "no"},
    )


# ── 5. Listing 批量改写 ─────────────────────────────────────────────────

@router.post("/listing-rewrite")
async def listing_rewrite(req: ListingRewriteReq) -> StreamingResponse:
    """批量 Listing 改写 — 多 ASIN 标题/五点/QA 批量生成。SSE 流式返回。"""
    from app.services import ai_synthesis_service

    asin_list = ", ".join(req.asins)
    fields_str = ", ".join(req.fields)

    prompt = f"""你是一位 Amazon Listing 文案专家。
请对以下 ASIN 进行批量文案改写：{asin_list}
站点：{req.marketplace}
改写字段：{fields_str}
改写风格：{req.style}

要求：
1. 先抓取每个 ASIN 的现有 Listing 内容
2. 分析各 ASIN 的卖点差异和目标受众
3. 针对每个 ASIN 逐一改写指定字段
4. 改写原则：
   - 标题：核心关键词前置，200字符内，包含品牌+核心卖点+关键属性
   - 五点：每点150-200字符，以利益点开头，融入关键词
   - 描述：情感化叙述，突出使用场景
   - QA：预测买家常见问题并给出专业回答
5. 输出格式：每个 ASIN 独立一节，包含改写前后对比

请用中文输出分析过程，英文输出改写文案。"""

    async def generator():
        start = time.time()
        yield _sse({"type": "phase", "phase": "rewriting"})
        try:
            async for prov, chunk in ai_synthesis_service.synthesize_native(
                "asin", asin_list, req.marketplace
            ):
                if prov == "_attempt":
                    yield _sse({"type": "attempt", "provider": chunk})
                elif prov == "error":
                    yield _sse({"type": "error", "detail": chunk})
                    return
                else:
                    yield _sse({"type": "token", "text": chunk, "provider": prov})
            elapsed = round(time.time() - start, 1)
            yield _sse({"type": "done", "provider": "hermes", "elapsed_s": elapsed})
        except Exception as exc:
            yield _sse({"type": "error", "detail": str(exc)})

    return StreamingResponse(
        generator(),
        media_type="text/event-stream",
        headers={"Cache-Control": "no-cache", "X-Accel-Buffering": "no"},
    )

File: client/src/api/deepAnalysis.ts
/* Deep Analysis API client */
import { api } from "./client";

export interface KeywordReq {
  keyword: string;
  country?: string;
  asin?: string;
}

export interface CompetitorReq {
  asin: string;
  country?: string;
  time_type?: string;
  time_value?: string;
}

export interface TrafficReq {
  asin: string;
  country?: string;
}

export interface ReviewsReq {
  asin: string;
  country?: string;
}

export interface ListingRewriteReq {
  asins: string[];
  marketplace?: string;
  fields?: string[];
  style?: string;
}

export async function keywordCompetition(req: KeywordReq) {
  const { data } = await api.post("/deep-analysis/keyword", req);
  return data;
}

export async function competitorLookup(req: CompetitorReq) {
  const { data } = await api.post("/deep-analysis/competitor", req);
  return data;
}

export async function trafficDiagnosis(req: TrafficReq) {
  const { data } = await api.post("/deep-analysis/traffic", req);
  return data;
}

/* SSE streaming for reviews and listing-rewrite */
export type SseEvent =
  | { type: "phase"; phase: string }
  | { type: "attempt"; provider: string }
  | { type: "token"; text: string; provider: string }
  | { type: "error"; detail: string }
  | { type: "done"; provider: string; elapsed_s: number };

export function streamReviews(
  req: ReviewsReq,
  onEvent: (evt: SseEvent) => void,
  signal?: AbortSignal,
): Promise<void> {
  return fetch("/api/deep-analysis/reviews", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    credentials: "include",
    body: JSON.stringify(req),
    signal,
  }).then((resp) => {
    if (!resp.ok) throw new Error(`HTTP ${resp.status}`);
    return consumeSSE(resp, onEvent);
  });
}

export function streamListingRewrite(
  req: ListingRewriteReq,
  onEvent: (evt: SseEvent) => void,
  signal?: AbortSignal,
): Promise<void> {
  return fetch("/api/deep-analysis/listing-rewrite", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    credentials: "include",
    body: JSON.stringify(req),
    signal,
  }).then((resp) => {
    if (!resp.ok) throw new Error(`HTTP ${resp.status}`);
    return consumeSSE(resp, onEvent);
  });
}

export async function consumeSSE(
  resp: Response,
  onEvent: (evt: SseEvent) => void,
): Promise<void> {
  const reader = resp.body?.getReader();
  if (!reader) return;
  const decoder = new TextDecoder();
  let buffer = "";
  while (true) {
    const { done, value } = await reader.read();
    if (done) break;
    buffer += decoder.decode(value, { stream: true });
    const lines = buffer.split("\n");
    buffer = lines.pop() || "";
    for (const line of lines) {
      const trimmed = line.trim();
      if (!trimmed || trimmed.startsWith(":")) continue;
      if (trimmed.startsWith("data:")) {
        try {
          const evt = JSON.parse(trimmed.slice(5).trim());
          onEvent(evt);
        } catch { /* ignore */ }
      }
    }
  }
}
